lib/new.js

```diff
@@ -0,0 +1,687 @@
+/**
+ * Copyright 2018 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the 'License');
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an 'AS IS' BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const path = require('path');
+const puppeteer = require('puppeteer-core');const findChrome = require('./find_chrome');
+const {rpc} = require('../rpc');
+const debugApp = require('debug')('carlo:app');
+const debugServer = require('debug')('carlo:server');
+const {Color} = require('./color');
+const {HttpRequest} = require('./http_request');
+
+const fs = require('fs');
+const util = require('util');
+const {URL} = require('url');
+const EventEmitter = require('events');
+const fsReadFile = util.promisify(fs.readFile);
+
+let testMode = false;
+
+class App extends EventEmitter {
+  /**
+   * @param {!Puppeteer.Browser} browser Puppeteer browser
+   * @param {!Object} options
+   */
+  constructor(browser, options) {
+    super();
+    this.browser_ = browser;
+    this.options_ = options;    this.windows_ = new Map();    this.exposedFunctions_ = [];
+    this.pendingWindows_ = new Map();
+    this.windowSeq_ = 0;
+    this.www_ = [];
+  }
+
+  async init_() {
```

Comment:
format: style mismatch:
```javascript
48|
49|  async init_() {
50|    debugApp('Configuring browser');
```
```suggestion
  async init_() {debugApp('Configuring browser');let page;await Promise.all([this.browser_.target().createCDPSession().then(session => {this.session_ = session;
```

⏎␣⁺␣⁺ at column 18 should be removed.
Triggered rule # 203
```
  +1.roles in {LITERAL} and not in {ARGUMENT}
	∧ +2.roles not in {RIGHT}
	∧ ^1.roles not in {OPERATOR, DECLARATION}
	∧ ^2.internal_type not in {BlockStatement}
	∧ -1.roles not in {STRING, EXPRESSION}
	∧ -1.reserved not in {,, :, .}
	∧ -2.diff_line = 0
	∧ -3.diff_line = 0
	∧ -3.label in {<newline>}
	∧ -4.roles not in {CALL}
	∧ -4.reserved not in {;}
	∧ -4.diff_offset ≥ 10
	∧ -5.label not in {<space>}
	⇒ y = '
	Confidence: 0.550. Support: 10.
```
⏎ at column 37 should be removed.
Triggered rule # 260
```
  +1.roles not in {ARGUMENT, CALLEE, LITERAL}
	∧ +1.reserved not in {=, {}
	∧ +2.length ≤ 49
	∧ +2.roles not in {RIGHT}
	∧ +3.roles not in {BLOCK}
	∧ ^1.roles not in {IDENTIFIER, OPERATOR, EXPRESSION, DECLARATION}
	∧ ^1.internal_type not in {BlockStatement}
	∧ -1.roles not in {STRING, EXPRESSION}
	∧ -1.reserved not in {,, :, ), .}
	∧ -1.diff_offset ≤ 14
	∧ -2.diff_line = 0
	∧ -2.roles not in {CALL}
	∧ -2.diff_col ≤ 8
	∧ -2.length ≥ 3
	∧ -2.reserved not in {(}
	∧ -2.label not in {<-space>}
	∧ -3.label not in {<newline>}
	∧ -4.diff_line = 0
	∧ -4.roles not in {CALL}
	∧ -4.reserved not in {;, {}
	∧ -4.label not in {<newline>}
	∧ -4.diff_col ≤ 10
	∧ -5.diff_offset ≥ 10
	∧ -5.label not in {<newline>}
	⇒ y = ⏎
	Confidence: 0.625. Support: 4.
```
⏎ at column 14 should be removed.
Triggered rule # 274
```
  +1.roles not in {ARGUMENT, CALLEE}
	∧ +1.reserved not in {=, {}
	∧ +2.roles not in {RIGHT}
	∧ +2.reserved = (
	∧ ^1.roles in {BLOCK} and not in {IDENTIFIER, OPERATOR, DECLARATION}
	∧ ^2.roles not in {CALL}
	∧ -1.roles not in {STRING, EXPRESSION}
	∧ -1.reserved not in {,, :, ), .}
	∧ -1.diff_offset ≤ 1
	∧ -2.diff_line = 0
	∧ -2.label not in {<-space>}
	∧ -2.diff_col ≤ 3
	∧ -3.label not in {<newline>}
	∧ -4.roles not in {CALL}
	∧ -4.reserved not in {;}
	⇒ y = ⏎
	Confidence: 0.837. Support: 46.
```
⏎␣⁺␣⁺ at column 24 should be removed.
Triggered rule # 214
```
  +1.roles in {CALLEE} and not in {ARGUMENT}
	∧ +2.roles not in {RIGHT}
	∧ ^1.roles not in {OPERATOR, DECLARATION}
	∧ ^2.roles in {ARGUMENT}
	∧ -1.roles not in {STRING, EXPRESSION}
	∧ -1.reserved not in {,, :, .}
	∧ -2.diff_line = 0
	∧ -3.label not in {<newline>}
	∧ -4.roles not in {CALL}
	∧ -4.reserved not in {;}
	∧ -4.diff_col ≤ 7
	⇒ y = ⏎
	Confidence: 0.994. Support: 88.
```
⏎␣⁺␣⁺ at column 66 should be removed.
Triggered rule # 122
```
  +1.roles in {ARGUMENT}
	∧ +2.roles not in {RIGHT}
	∧ +4.reserved not in {const}
	∧ ^1.roles not in {DECLARATION}
	∧ -1.roles not in {STRING, EXPRESSION}
	∧ -1.reserved not in {:, .}
	∧ -2.diff_line = 0
	∧ -3.length ≤ 8
	∧ -3.diff_offset ≥ 4
	∧ -3.label in {<space>}
	∧ -4.reserved = =
	⇒ y = '
	Confidence: 0.500. Support: 19.
```